# Preliminary Steps

In [84]:
import pandas as pd
import numpy as np
import re
import os

In [85]:
# Download and unzip shivamb/netflix-shows dataset
!kaggle datasets download shivamb/netflix-shows 

Dataset URL: https://www.kaggle.com/datasets/shivamb/netflix-shows
License(s): CC0-1.0
netflix-shows.zip: Skipping, found more recently modified local copy (use --force to force download)


In [86]:
import subprocess

In [87]:
# Only unzip if netflix-shows.zip hasn't been unzipped already
dir_files = os.listdir(os.getcwd())
unzipped_csvs = ['netflix_titles.csv']

if set(dir_files).isdisjoint(set(unzipped_csvs)) and "netflix-shows.zip" in dir_files:
    import zipfile
    with zipfile.ZipFile("netflix-shows.zip", "r") as zip_ref:
        zip_ref.extractall()

Try reading in the data on your own!

In [88]:
# TODO: Read your data as a pandas dataframe. Save the dataframe to a variable named "netflix"
DATA_PATH = os.path.join(os.getcwd(), "netflix_titles.csv")
netflix =netflix = pd.read_csv(DATA_PATH)
netflix

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

# EDA

We have a much larger dataset than last week. What are all the columns and what do they mean? What does each row mean? Take a some time to look through the data and understand what we're working with.

Some useful functions: describe(), unique(), sort_values(), dtypes(), shape, columns, info(), isnull(), value_counts(), and more if you can think of them. Get used to exploring data like this so you know what you're working with first!

In [89]:
netflix["type"].value_counts()
netflix["release_year"].value_counts().head()
netflix.info()
netflix.isna().sum()
netflix.dropna()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


show_id     type                          title             director  \
7         s8    Movie                        Sankofa         Haile Gerima   
8         s9  TV Show  The Great British Baking Show      Andy Devonshire   
9        s10    Movie                   The Starling       Theodore Melfi   
12       s13    Movie                   Je Suis Karl  Christian Schwochow   
24       s25    Movie                          Jeans           S. Shankar   
...      ...      ...                            ...                  ...   
8801   s8802    Movie                        Zinzana      Majid Al Ansari   
8802   s8803    Movie                         Zodiac        David Fincher   
8804   s8805    Movie                     Zombieland      Ruben Fleischer   
8805   s8806    Movie                           Zoom         Peter Hewitt   
8806   s8807    Movie                         Zubaan          Mozez Singh   

                                                   cast  \
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8     Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
24    Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
8                                        United Kingdom  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
24                                                India  September 21, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating   duration  \
7             1993  TV-MA    125 min   
8             2021  TV-14  9 Seasons   
9             2021  PG-13    104 min   
12            2021  TV-MA    127 min   
24            1998  TV-14    166 min   
...            ...    ...        ...   
8801          2015  TV-MA     96 min   
8802          2007      R    158 min   
8804          2009      R     88 min   
8805          2006     PG     88 min   
8806          2015  TV-14    111 min   

                                             listed_in  \
7     Dramas, Independent Movies, International Movies   
8                         British TV Shows, Reality TV   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
24     Comedies, International Movies, Romantic Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  
7     On a photo shoot in Ghana, an American model s...  
8     A talented batch of amateur bakers face off in...  
9     A woman adjusting to life after a loss contend...  
12    After 

# Granularity, Scope, and Temporality

Once you have completed your EDA, you should be able to answer each of these broad ideas about the data. If not, you can always do more EDA!

*Note, you can jot down some quick notes instead of a long answer to help remind yourself about certain characteristics of this dataframe

#### Granularity: What information does each row give us? How is each row unique (i.e., what identifies each row? This is called the **"primary key"**)

The dataset is giving us information about different tv shows in movies, the primary key is the show_id

#### Scope: What are some interesting things we could learn from the data (we can find this out by exploring the columns we have)? Do we have to manipulate the data in some way to get what we're interested in?

We can learn which years have the most release dates, which directors have done a lot of work, which countries have produced the most Netflix content, and the range of ratings of each piece of entertainment

#### Temporality: When was the data collected? How often is the data collected (if there is a pattern)? Do we need to adjust for consistency in the dates/times? Do we need to adjust the data type of our time variables?

The dataset includes temporal information through columns such as release_year and data_added. These tell us when content was released and when it was added to Netflix. The data is probably updated periodically, to analyze time trends we can use datatime type so we can filter by year, month, or time ranges.

We will be having people share their responses to each of these questions and discuss what everybody found in the data!

# Faithfulness

This is all about deciding whether you can trust the data in the form it came in, or whether you need to make adjustments to do so.

Did you find any strange or inconsistent values? Can you figure out how the data was collected? Are there any duplicate values (in this case, there shouldn't be because each movie is a separate one)?

I found that there was a lot of missing directors from the data, as well as country and cast. Most likely, that information wasn't public when the dataset was made. I didn't see any obvious impossible values or duplicate values. The data was probably collected from Netlix's database and done automatically.

Possibly one of the most important questions is what to do with default values. These can be NaN, NA, 0, or something else entirely. Based on what we see from this data, what should we do with our missing values, and should we take different steps for different columns? You can check the slides for some examples of what people do with missing data.

There are several missing values, but we shouldn't treat them all the same because they are different types of information. For example, a missing value in the director column is not the same as a missing type in the country column. We could address it by just dropping them or by adding something like "Unknown". 

We'll discuss again afterwards!

# Variable Typing

Just for practice, it's important to be able to identify what type of variable something is just by looking at it. As a refresher, the 4 types are Quantitative Continous, Quantitative Discrete, Qualitative Ordinal, and Qualitative Nominal.

#### Question: What are the variables types of
a) Director

b) Country

c) Date Added

d) Release Year


a) Qualitative Nominal
b) Qualitative Nominal
c) Quantitative Continuous
d) Quantitative Discrete

# Main Exercise

A common, but often misguided, way to handle datasets that have outliers/untrustworthy data is to drop all rows or columns that have missing values.


Find the shape of our dataset before we do any data manipulation and compare it with the shape after you use the dropna function. What do you notice and how many columns or rows were removed from our data?

Tip: You should make a deep copy of a dataframe by using {data name}.copy(deep=True). Give it an appropriate name to reflect the fact that this will be the dataframe that will have all null values dropped without any previous operations performed on it.

In [90]:
print (netflix.shape)
netflix_no_nulls = netflix.copy(deep=True)
netflix_no_nulls = netflix_no_nulls.dropna()
print (netflix_no_nulls.shape)

(8807, 12)
(5332, 12)


You should've found that some of our data was removed after using the drop_na function. Let's go back to our original dataset before we used the function.

Try creating a list of the columns that have null values. Feel free to search things up on StackOverflow, the Pandas documentations, etc.

In [91]:
cols_with_nulls = netflix.columns[netflix.isnull().any()]
print (cols_with_nulls)

Index(['director', 'cast', 'country', 'date_added', 'rating', 'duration'], dtype='object')


Now that we have the columns where missing values are present, we have a better idea of what the data types of our missing values actually area.

Try filling in the missing director values with an empty string "". You can check if you did this operation correctly by checking the columns that have missing values and replacing the original dataframe name in the code with the name of the copy you should make for this task.

In [92]:
netflix_copy = netflix.copy (deep=True)
netflix_copy ["director"] = netflix_copy ["director"].fillna ("")
netflix_copy ["director"]

0       Kirsten Johnson
1                      
2       Julien Leclercq
3                      
4                      
             ...       
8802      David Fincher
8803                   
8804    Ruben Fleischer
8805       Peter Hewitt
8806        Mozez Singh
Name: director, Length: 8807, dtype: object

Let's check if filling the "director" column will help reduce the number of rows that will be removed when we use the dropna function.

Use the dropna function on the dataframe that is created after filling in the "director" column and find its shape. Compare the shape of this dataframe to the shape of the original dataframe and the dataframe after dropping all null values.

In [93]:
print (netflix.shape)
print (netflix_copy.dropna().shape)
print (netflix.dropna().shape)

(8807, 12)
(7290, 12)
(5332, 12)


You should notice that the size of the dataframe that is created when we drop null values after filling in the "director" column will have less rows than the original dataframe, but more than the dataframe that is created when we just drop all values without performing any operations.

This should serve as a basic exercise on how you can approach data that is not 100% clean. Most data in consulting projects and in the real world will force you to find ways to balance the preservation of the original data and getting rid of the unecessary parts of the data.

### String Manipulation. We'll load in a new dataset for this part.

In [94]:
import os
print(os.getcwd())
print(os.listdir())

C:\Users\jwags\JCP26Notebooks
['.git', '.ipynb_checkpoints', 'netflix-shows.zip', 'netflix_titles.csv', 'README.md', 'sentiment140.zip', 'Week_1_Pandas.ipynb', 'Week_2_EDA_DataCleaning_JCP_Copy.ipynb']


In [103]:
import pandas as pd
import zipfile
import os

with zipfile.ZipFile('sentiment140.zip', 'r') as z:
    z.extractall()

tweets_data = pd.read_csv('training.1600000.processed.noemoticon.csv', 
                          encoding='latin-1', 
                          header=None,
                          names=['Polarity', 'id', 'Date', 'NO_QUERY', 'Username', 'Tweet'])
print(tweets_data.shape)
print(tweets_data.head())

(1600000, 6)
   Polarity          id                          Date  NO_QUERY  \
0         0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY   
1         0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   
2         0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
3         0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4         0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   

          Username                                              Tweet  
0  _TheSpecialOne_  @switchfoot http://twitpic.com/2y1zl - Awww, t...  
1    scotthamilton  is upset that he can't update his Facebook by ...  
2         mattycus  @Kenichan I dived many times for the ball. Man...  
3          ElleCTF    my whole body feels itchy and like its on fire   
4           Karoli  @nationwideclass no, it's not behaving at all....  


In [104]:
tweets = tweets_data.drop(columns = ["NO_QUERY"]).rename(columns={"Polarity": "Polarity",
                                                                  "id": "id",
                                                                  "Date": "Date",
                                                                  "Username" : "Username"})
final_tweets = tweets.set_axis([*tweets.columns[:-1], 'Tweet'], axis=1)
final_tweets.head(5)

Polarity          id                          Date         Username  \
0         0  1467810369  Mon Apr 06 22:19:45 PDT 2009  _TheSpecialOne_   
1         0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton   
2         0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
3         0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ElleCTF   
4         0  1467811193  Mon Apr 06 22:19:57 PDT 2009           Karoli   

                                               Tweet  
0  @switchfoot http://twitpic.com/2y1zl - Awww, t...  
1  is upset that he can't update his Facebook by ...  
2  @Kenichan I dived many times for the ball. Man...  
3    my whole body feels itchy and like its on fire   
4  @nationwideclass no, it's not behaving at all....

#### Now that we have our **final_tweets** dataset, we can start working with it. Let's practice string methods first.

In [106]:
final_tweets["Username"].str.lower()

0          _thespecialone_
1            scotthamilton
2                 mattycus
3                  ellectf
4                   karoli
                ...       
1599995    amandamarie1028
1599996        thewdboards
1599997             bpbabe
1599998       tinydiamondz
1599999     ryantrevmorris
Name: Username, Length: 1600000, dtype: object

#### Try making all the entries in the Tweet column uppercase and save this in a final_tweets_upper dataframe.

In [107]:
final_tweets_copy = final_tweets.copy (deep=True)
final_tweets_copy["Tweet"].str.upper()

0          @SWITCHFOOT HTTP://TWITPIC.COM/2Y1ZL - AWWW, T...
1          IS UPSET THAT HE CAN'T UPDATE HIS FACEBOOK BY ...
2          @KENICHAN I DIVED MANY TIMES FOR THE BALL. MAN...
3            MY WHOLE BODY FEELS ITCHY AND LIKE ITS ON FIRE 
4          @NATIONWIDECLASS NO, IT'S NOT BEHAVING AT ALL....
                                 ...                        
1599995    JUST WOKE UP. HAVING NO SCHOOL IS THE BEST FEE...
1599996    THEWDB.COM - VERY COOL TO HEAR OLD WALT INTERV...
1599997    ARE YOU READY FOR YOUR MOJO MAKEOVER? ASK ME F...
1599998    HAPPY 38TH BIRTHDAY TO MY BOO OF ALLL TIME!!! ...
1599999    HAPPY #CHARITYTUESDAY @THENSPCC @SPARKSCHARITY...
Name: Tweet, Length: 1600000, dtype: object

#### Now let's try to replace something in a string!

In [111]:
final_tweets[final_tweets['Tweet'].str.contains('_')]

Polarity          id                          Date         Username  \
8               0  1467811795  Mon Apr 06 22:20:05 PDT 2009  2Hood4Hollywood   
22              0  1467814119  Mon Apr 06 22:20:40 PDT 2009        cooliodoc   
69              0  1467825084  Mon Apr 06 22:23:30 PDT 2009    PresidentSnow   
109             0  1467838188  Mon Apr 06 22:26:54 PDT 2009      jess_higley   
119             0  1467839586  Mon Apr 06 22:27:18 PDT 2009        sonyolmos   
...           ...         ...                           ...              ...   
1599987         4  2193579092  Tue Jun 16 08:38:58 PDT 2009        cathriiin   
1599988         4  2193579191  Tue Jun 16 08:38:59 PDT 2009          tellman   
1599989         4  2193579211  Tue Jun 16 08:38:59 PDT 2009        jazzstixx   
1599993         4  2193579477  Tue Jun 16 08:39:00 PDT 2009      ChloeAmisha   
1599994         4  2193579489  Tue Jun 16 08:39:00 PDT 2009        EvolveTom   

                                                     Tweet  
8                     @Tatiana_K nope they didn't have it   
22        @angry_barista I baked you a cake but I ated it   
69       @Lt_Algonquin agreed, I saw the failwhale alll...  
109      @marykatherine_q i know! I heard it this after...  
119      @eRRe_sC aaw i miss ya all too.. im leaving to...  
...                                                    ...  
1599987  @La_r_a NEVEER  I think that you both will get...  
1599988  @Roy_Everitt ha- good job. that's right - we g...  
1599989                 @Ms_Hip_Hop im glad ur doing well   
1599993                                  @SCOOBY_GRITBOYS   
1599994  @Cliff_Forster Yeah, that does work better tha...  

[75329 rows x 5 columns]

Task 2: In the cell below, use .str.replace to replace all the underscores in Username with a space (" ") in the tweeters_with__ dataframe. Output the changed tweeters_with__ dataframe.

Doc: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html

In [115]:
final_tweets["Username"].replace ("_", "")

0          _TheSpecialOne_
1            scotthamilton
2                 mattycus
3                  ElleCTF
4                   Karoli
                ...       
1599995    AmandaMarie1028
1599996        TheWDBoards
1599997             bpbabe
1599998       tinydiamondz
1599999     RyanTrevMorris
Name: Username, Length: 1600000, dtype: object

# Regex

---
### Part 1 — Does the title start with a number? 

Some Netflix titles start with a number, like `"13 Reasons Why"` or `"21 Jump Street"`.

**Task:** Use `str.contains()` with a regex pattern to find all titles that **start with a digit**. Store the result in a new DataFrame called `number_titles` and print how many there are.

**Hint:** Should be 130


In [117]:
# TODO: write a pattern that matches titles starting with a digit
pattern = r"^\d"
number_titles = netflix[netflix["title"].str.contains(pattern, na=False)]
print(f"Titles starting with a digit: {len(number_titles)}")
number_titles["title"].head(10)

Titles starting with a digit: 130


188                 2 Alone in Paris
323                          30 Rock
324                          44 Cats
404    9to5: The Story of a Movement
438                 2 Weeks in Lagos
558                        6 Bullets
774                         2 Hearts
850                         99 Songs
851                 99 Songs (Tamil)
852                99 Songs (Telugu)
Name: title, dtype: object

---
### Part 2 — Search descriptions for key words using `\w` and `[ ]`

**Task:** Find all titles whose description mentions the word `"love"` **or** `"romance"` (case-insensitive). Count them and print 5 examples.

Then, as a follow-up, find titles whose description contains **a number followed immediately by the word `"day"` or `"days"`** — e.g. `"30 days"` or `"1 day"`.

**Hints:**
- Pass `case=False` to `str.contains()` to ignore capitalisation.
- For the number + day pattern: `\d+` matches the number, `\s` matches the space, `days?` matches `"day"` or `"days"` (the `?` makes the `s` optional).
- Answer should be 826 and 9 respectively


In [118]:
# TODO: descriptions mentioning love or romance (case-insensitive)
pattern_love = r"love|romance"  # <-- your pattern here

love_titles = netflix[netflix["description"].str.contains(pattern_love, case=False, na=False)]
print(f"Titles mentioning love or romance: {len(love_titles)}")
love_titles["title"].head(5)

Titles mentioning love or romance: 826


24                   Jeans
25    Love on the Spectrum
26          Minsara Kanavu
27               Grown Ups
30         Ankahi Kahaniya
Name: title, dtype: object

In [122]:
# TODO: descriptions mentioning a number followed by "day" or "days"
pattern_days = r"\d+ days?" 

days_titles = netflix[netflix["description"].str.contains(pattern_days, case=False, na=False)]
print(f"Titles with 'N day(s)' in description: {len(days_titles)}")
days_titles[["title", "description"]].head(5)

Titles with 'N day(s)' in description: 9


title  \
1565                         Just The Way You Are   
2888                                Hi Bye, Mama!   
4829                             Sunday's Illness   
5031                                    Forgotten   
5893  Winter on Fire: Ukraine's Fight for Freedom   

                                            description  
1565  An overconfident teen bets he can make a homel...  
2888  When the ghost of a woman gains a second chanc...  
4829  Decades after being abandoned as a young child...  
5031  When his abducted brother returns seemingly a ...  
5893  Over 93 days in Ukraine, what started as peace...